homeassistant/components/tesla/config_flow.py

```diff
@@ -140,6 +140,8 @@ async def validate_input(hass: core.HomeAssistant, data):
         (config[CONF_TOKEN], config[CONF_ACCESS_TOKEN]) = await controller.connect(
             test_login=True
         )
+        config[CONF_USERNAME] = data[CONF_USERNAME]
```

Comment:
It's a significant rewrite and everyone is locked out now. That was the next thing to do

Reply:
Yeah, but we don't accept web scraping in integrations, so we'll probably not accept this fix.

Reply:
As much as I hate to say it, this integration should probably be moved to `hacs` unless there is a way around the scraping. https://tesla-api.timdorr.com/api-basics/authentication#step-2-obtain-an-authorization-code hasn't seemed to find a way around it yet.

Reply:
If we use an oauth2 flow the authentication will be done by the user in the browser. We don't need to simulate the user I think. I'm looking from the outside of course.

Reply:
It's doable but it will take a rewrite of the HA component as there's no logic to detect a need to reauth.  Right now, it relies on automatic reauth. If you take away the webscraping, then every reboot will require the user remove and readd because we don't have any logic in place to detect and provide a reauth. I may not have time to do that in the short term.

Reply:
As an update, external oauth won't work it appears. Tesla is checking the redirect_uri and stop the request if it isn't the Tesla one.

Reply:
That's unfortunate. I've asked for a second opinion from other core members.

Reply:
@MartinHjelmare You may want to talk to @balloob too. I discussed this with him yesterday when I first discovered the issue.

Reply:
Yes, he's in the loop.

Reply:
Following. MyQ integration with new OAUTH API has same issue where web scraping is required to get values that have to be returned as part of authentication